chore: Update accounts controller v14

tommasini opened the pull request:
## **Description**

This PR updates accounts controller to v^14.

qa build: https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/96496314-9441-4329-bd43-fc9e4f1e3dcd

This PR also includes this PR: https://github.com/MetaMask/metamask-mobile/pull/9832

## **Related issues**

## Blocked by

- [x] https://github.com/MetaMask/metamask-mobile/pull/9718
  This PR is dependent on the snaps controller update [PR](https://github.com/MetaMask/metamask-mobile/pull/9718)
- [x] https://github.com/MetaMask/metamask-mobile/pull/9570
  This PR is dependent on the keyring controller update [PR](https://github.com/MetaMask/metamask-mobile/pull/9570)

## Fixes
- https://github.com/MetaMask/mobile-planning/issues/1777

## **Manual testing steps**

1. Import account via SRP
2. Import account via private key
3. Import tokens 

## **Screenshots/Recordings**

* Import with private key

![image](https://github.com/MetaMask/metamask-mobile/assets/46944231/fdfe0e97-bfc8-4a62-ae76-3aebc91fd8fe)


* Importing with SRP


https://github.com/MetaMask/metamask-mobile/assets/46944231/d829d4d6-6e28-404f-b1bb-e0ed3ec6f986



### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

tommasini committed:
Update accounts controller v14, all breaking changes and patches addressed, missing typescript issues

tommasini committed:
merge main and solve conflicts

tommasini committed:
fix lint ts issues on test files

tommasini committed:
fix lint and dedupe issues

tommasini committed:
fix typescript issue

tommasini committed:
Merge branch 'main' into chore/update-accounts-controller-v14

owencraston commented on app/util/test/accountsControllerTestUtils.ts:
```diff
@@ -59,21 +58,23 @@ export const internalAccount2 = createMockInternalAccount(
+};
+
 // used as a default mock for other tests
-export const MOCK_ACCOUNTS_CONTROLLER_STATE: AccountsControllerState = {
```
Nitpick: I want to keep these type explicit in case there is an update in the data type that we expect form the package we can find the issue very easily before it fails on CI.

tommasini replied:
Reverted <3

owencraston commented:
we should also bump the version of the [keyring-api](https://github.com/MetaMask/keyring-api/releases). I think we will need [a minimum of 6.1.1](https://github.com/MetaMask/core/pull/4262/files#diff-11700fa807e8d925b649057d40f964047063e26926ce62fc5f09210de947dbffR47). This update introduces an extra field on the InternalAccount type of importTime. This will allow us to 100% migrate off of preferences controller.

tommasini committed:
change metadata and bump keyring api to have same types

tommasini committed:
revert unnecessary object on accounts controller test utils

owencraston commented on package.json:
```diff
@@ -149,7 +149,7 @@
     "@metamask/gas-fee-controller": "^13.0.0",
     "@metamask/key-tree": "^9.0.0",
-    "@metamask/keyring-api": "^4.0.0",
+    "@metamask/keyring-api": "^6.0.0",
```
I believe version 14 of the accounts controller is using `^6.1.1`

tommasini replied:
Tried to bump because it was not a major version change, but it seems it is ^6.0.0: https://github.com/MetaMask/core/blob/3227bd51bc4eca0531a53f4be6468404e6a5f032/packages/accounts-controller/package.json#L47

tommasini committed:
update importTime for real values

tommasini committed:
unit tests fixed with hardcoded date value

tommasini commented on app/store/migrations/036.ts:
```diff
@@ -115,6 +115,7 @@ async function createInternalAccountsForAccountsController(
       options: {},
       metadata: {
         name: identity.name,
+        importTime: 123,
```
Addressed! You are super right! Thanks for the catch!

montelaidev reviewed (changes requested):
We will need to create a new migration to add `importTime` to existing users that have already migrated past `036`. 

An alternative would be to call `updateAccounts` on the accounts controller and it will populate the value.

legobeat commented on yarn.lock:
```diff
@@ -3633,6 +3633,23 @@
     immer "^9.0.6"
     uuid "^8.3.2"
 
+"@metamask/accounts-controller@^14.0.0":
```
v11 is still used by `@metamask/assets-controllers`.

I guess this also needs to be coupled with an update of `@metamask/assets-controllers` to either v29 or v30?

```
$ nvs -p @metamask/accounts-controller @metamask/assets-controllers | tail
26.0.0  2024-03-04T18:44:07.058Z        >=16.0.0        ^11.0.0
27.0.0  2024-03-12T19:47:40.443Z        >=16.0.0        ^12.0.0
27.0.1  2024-03-13T11:26:43.777Z        >=16.0.0        ^12.0.0
27.1.0  2024-03-19T08:49:30.180Z        >=16.0.0        ^12.0.0
27.2.0  2024-03-19T18:15:55.457Z        >=16.0.0        ^12.0.0
28.0.0  2024-04-17T15:00:35.910Z        >=16.0.0        ^13.0.0
29.0.0  2024-05-01T13:38:25.311Z        >=16.0.0        ^14.0.0
30.0.0  2024-05-29T20:08:33.688Z        >=16.0.0        ^14.0.0
31.0.0  2024-05-31T18:08:03.582Z        >=16.0.0        ^15.0.0
32.0.0  2024-05-31T19:50:52.965Z        ^18.18 || >=20  ^16.0.0
```

tommasini replied:
But since it's a dependency of assets controller and not a peer dependency, is this a blocker?

Assets controllers will still be using the accounts controller v11, do you think it should be added as a peer dependency instead?

legobeat replied:
The dependency is only referenced in `TokenDetectionController` Looking here:
https://github.com/MetaMask/core/blob/main/packages/assets-controllers/src/TokenDetectionController.ts

The actual code of the dependency is not actually used - it's just pulled in for types and interacting with an externally provided `AccountsController`.

So unless we have them matched here, we will have an `assets-controller` implementet towards and expecting an `accounts-controller@11` but actually getting `accounts-controller@14`.

If it has been verified that there are no compatibility issues with mixing v11/v14.

(

So yes, I do think it should be in `peerDependencies`. At which point the `dependencies` entry can be removed, as the expected types will be provided through the `peerDependencies`.

tommasini replied:
Thanks for this answer and thanks for validating that there are no compatibility issues!
With this being said I'm WIP on the [assets controllers v30 update ](https://github.com/MetaMask/metamask-mobile/pull/9832)as we speak and that [version](https://github.com/MetaMask/core/blob/6f563ffb714db410666ec518e76e1806581367c8/packages/assets-controllers/package.json#L91) already have accounts-controller as a peer dependency

montelaidev commented:
> We will need to create a new migration to add `importTime` to existing users that have already migrated past `036`.
> 
> An alternative would be to call `updateAccounts` on the accounts controller and it will populate the value.

Also, i think the importTime from the identity in the preferences controller should be used if its available

tommasini committed:
add migrations for imporTime property on accounts controller internal accounts acconts metadata

tommasini committed:
remove unused variable

tommasini commented:
> Also, i think the importTime from the identity in the preferences controller should be used if its available

Ah read that right now, will update the migration with it

tommasini committed:
update migration with the import time of internal accounts be the importTime on the identities of preferences controller if it exists if not it fallsback to date now

tommasini committed:
Merge branch 'main' into chore/update-accounts-controller-v14

tommasini committed:
merge main and solve conflicts

tommasini committed:
fix lint issues

tommasini committed:
Merge branch 'main' into chore/update-accounts-controller-v14

montelaidev commented on app/store/migrations/043.test.ts:
```diff
@@ -0,0 +1,199 @@
+      expect(
+        newState.engine.backgroundState.AccountsController.internalAccounts
+          .accounts[accountId].metadata.importTime,
+      ).toBeDefined();
```
We can use `expect.any(Number)` here

tommasini commented on app/store/migrations/043.test.ts:
```diff
@@ -0,0 +1,199 @@
+      oldState,
+    ) as Partial<RootState>;
+    if (!newState.engine) {
+      expect(true).toBeFalsy();
```
Was a workaround that I've found to the test fail if the engine is not defined and to deal with the lint and typescript issues to check if the engine object was defined inside of newState, since it can be undefined. It will never happen!

tommasini committed:
update test to expect any Number

tommasini commented on app/store/migrations/043.ts:
```diff
@@ -0,0 +1,110 @@
+        !accountsControllerState.internalAccounts.accounts[accountId].metadata
+          .importTime
+      ) {
+        Object.keys(preferencesControllerState.identities).forEach(
```
I'm wondering if there is any reason for the identities to be empty objects and if we should take this into account and write a fallback if it isn't, it seems safer but at the same time, it seems dead code if it's not possible such scenario

tommasini replied:
This outdated because the 043 file represents other migration but this is still present on 044.ts file

montelaidev replied:
Yes we should still check because its possible for the user to reset their vault via forget password after running migration 036. In this case the identities should be an empty object.

tommasini replied:
addressed on this [commit](https://github.com/MetaMask/metamask-mobile/pull/9779/commits/febbb833db08dd5cdbe2906df9919742657d5e0e)

tommasini committed:
first commit to reduce

tommasini committed:
merge main and solve conflicts

tommasini committed:
Merge branch 'main' into chore/update-accounts-controller-v14

legobeat committed:
chore: yarn deduplicate

legobeat committed:
Merge branch 'main' into chore/update-accounts-controller-v14

tommasini committed:
Merge branch 'main' into chore/update-accounts-controller-v14

montelaidev commented on app/store/migrations/044.test.ts:
```diff
@@ -0,0 +1,199 @@
+      expect(true).toBeFalsy();
+      return;
+    }
+    Object.keys(
```
```suggestion
    const newState: Pick<RootState, 'engine'> = migration(oldState) as Pick<
      RootState,
      'engine'
    >;
```

I think we can just use pick instead of partial and then we don't need to this ```
if (!newState.engine) {
      expect(true).toBeFalsy();
      return;
    }
```

tommasini replied:
You are right!  Thanks for the knowledge!

tommasini replied:
addressed on this [commit](https://github.com/MetaMask/metamask-mobile/pull/9779/commits/febbb833db08dd5cdbe2906df9919742657d5e0e)

tommasini committed:
address when identities are empty on the migration fallback importTime to have the current date as well and add test scenario

tommasini committed:
rename migrations to avoid conflicts when merging main

tommasini committed:
merge main and solve conflicts

tommasini committed:
update migration with better type safety protection and update test cases

tommasini committed:
remove cast on test file for migration 044

tommasini committed:
fix typescript and lint issues

tommasini committed:
migration 47

tommasini committed:
mrege main and solve conflicts

tommasini committed:
fix migration index

tommasini committed:
change test cases to use type of on migration of accounts controller

tommasini committed:
merge main and solve conflicts

tommasini committed:
fix typing issues on migration 42

tommasini committed:
Merge branch 'main' into chore/update-accounts-controller-v14

tommasini committed:
Merge branch 'main' into chore/update-accounts-controller-v14

tommasini committed:
Merge branch 'main' into chore/update-accounts-controller-v14

tommasini committed:
Merge branch 'main' into chore/update-accounts-controller-v14

Cal-L reviewed (dismissed):
LGTM

tommasini commented on yarn.lock:
```diff
@@ -3573,6 +3573,23 @@
     immer "^9.0.6"
     uuid "^8.3.2"
 
+"@metamask/accounts-controller@^14.0.0":
```
I can merge this PR into the assets controllers one: https://github.com/MetaMask/metamask-mobile/pull/9832
Or I can merge both on the same day, resolving by changing on the assets controllers one the migration for number 48*
I would like to keep those separate for now to make life easier for the reviewers on the assets controllers one

tommasini replied:
Pointed https://github.com/MetaMask/metamask-mobile/pull/9832 to this branch so we can proceed with easier reviews and at the same time keep this ones that we already have! Thanks for the suggestions Lego!

tommasini committed:
Merge branch 'main' into chore/update-accounts-controller-v14

tommasini committed:
chore: Update controller utils to v^10 and assets controllers utils to v^30 (#9832)

## **Description**

Asset scontrollers update for v^30
patch updated accordingly the core branch
`patch/mobile-assets-controllers-v30`

With the assets controllers updates there were breaking changes at the
TokenRatesController data structure, that replaces the
`contractExchangeRates` and `contractExchangeRatesByChainId ` properties
for a property called `marketData`

Controller utils patch removed and controller utils updated to v^10

we are removing the script to do patches automatically with core
repository-built code, since the size of the patch doesn't seem to
justify this and I believe we should work on the direction to obliterate
it from now on since we are close to the latest version of assets
controllers.


This PR is pointing to the [accounts controller update
PR](https://github.com/MetaMask/metamask-mobile/pull/9779) since they
need to be merged into main

## **Related issues**

Fixes: https://github.com/MetaMask/mobile-planning/issues/1783
https://github.com/MetaMask/mobile-planning/issues/1780

## **Manual testing steps**

1. Import wallet with tokens & funds
2. Import tokens via auto detection, custom , via search and via dapp
3. import nfts with auto detection, via custom, with display nft media
toggle on and off and ipfs gateway enabled on and off
4. Perform transactions & swaps to make sure the conversion to fiat is
working properly

## **Screenshots/Recordings**
Tokens:

https://github.com/MetaMask/metamask-mobile/assets/46944231/d3246625-6e70-4896-add3-dd92eb0bf8bb

Nfts:


https://github.com/MetaMask/metamask-mobile/assets/46944231/e6a642ee-72a1-4b8c-ad3b-4946d9068ac3



### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [x] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

---------

Co-authored-by: Nico MASSART <NicolasMassart@users.noreply.github.com>

tommasini committed:
Merge branch 'main' into chore/update-accounts-controller-v14

sethkfman reviewed (approved):
LGTM
